pkg/shipyard/cluster.go

```diff
@@ -0,0 +1,207 @@
+package shipyard
+
+import (
+	"fmt"
+	"net/http"
+	"os/exec"
+	"path/filepath"
+	"time"
+)
+
+const (
+	startupTimeout = 10 * time.Second
+)
+
+// Cluster encapsulates a mock Kubernetes cluster.
+type Cluster struct {
+	Options
+	Docker Docker
+
+	containers struct {
+		etcd    string
+		api     string
+		kubelet string
+	}
+
+	manifestDir   string
+	varLibDocker  string
+	varLibKubelet string
+	varRun        string
+}
+
+// SetUp the e2e cluster.
+func (cl *Cluster) SetUp() {
+	Log.Logf("SetUp")
+
+	cl.resolveDirs()
+	cl.pullImages()
+
+	cl.StartEtcd()
+	cl.StartApiServer()
+	cl.StartKubelet()
+
+	cl.WaitForApiServer()
+}
+
+// TearDown the e2e cluster.
+func (cl *Cluster) TearDown() {
+	Log.Logf("Teardown")
+
+	cl.StopKubelet()
+	cl.StopApiServer()
+	cl.StopEtcd()
+}
+
+func (cl *Cluster) resolveDirs() {
+	// TODO: directories should be configurable, but there seem to be issues with the
+	// the nsenter mounter that prevent us from moving the location of /var/lib/kubelet.
+	cl.manifestDir = fmt.Sprintf("%v/test/e2e/cluster/manifests", cl.BaseDir)
+	cl.varLibDocker = "/var/lib/docker"
+	cl.varLibKubelet = "/var/lib/kubelet"
+	cl.varRun = "/var/run"
+
+	var err error
+
+	cl.manifestDir, err = filepath.Abs(cl.manifestDir)
+	if err != nil {
+		Log.Fatal(err)
```

Comment:
I don't like there are so many places where the code just exits.

I think the intended use of the repo is a library. I'd like to see errors instead of immediate exiting.

Reply:
Makes total sense, on it